internal/metrics/info.go

```diff
@@ -0,0 +1,150 @@
+package metrics // import "github.com/pomerium/pomerium/internal/metrics"
+
+import (
+	"context"
+	"runtime"
+	"time"
+
+	"github.com/pomerium/pomerium/internal/version"
+
+	"github.com/pomerium/pomerium/internal/log"
+	"go.opencensus.io/metric"
+	"go.opencensus.io/metric/metricdata"
+	"go.opencensus.io/metric/metricproducer"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+var (
+	//buildInfo               = stats.Int64("build_info", "Build Metadata", "1")
+	configLastReload        = stats.Int64("config_last_reload_success_timestamp", "Timestamp of last successful config reload", "seconds")
+	configLastReloadSuccess = stats.Int64("config_last_reload_success", "Returns 1 if last reload was successful", "1")
+
+	// Metrics that may/must be measured or set up before the config is fully parsed
+	// and metric views are initialized
+	buildInfo      *metric.Int64Gauge
+	policyCount    *metric.Int64DerivedGauge
+	configChecksum *metric.Int64Gauge
```

Comment:
Global state always makes me a little nervous. 

What do you think about wrapping these metrics in a struct and protecting access via a mutex? It could also potentially help with abstracting out the nil checks and initial population. It feels like we want these to be [initialized global singletons](http://marcio.io/2015/07/singleton-pattern-in-go/)?  

Also, how would multiple instance effect the recording of these stats?

Reply:
>What do you think about wrapping these metrics in a struct and protecting access via a mutex?

I considered a struct for DRYness, but discounted it.  It didn't seem like it would become any less verbose with the _current_ metric count.  We have two type signatures and three methods.

Good point on access safety.  These are currently early single threaded calls, but that isn't guaranteed later.  This implementation is not safe.   

A lighter pattern might be to do the setup in init(), which I believe is [threadsafe](https://golang.org/ref/spec#Package_initialization), though possibly controversial.  Given the rest of the "metric" type variables are initialized in package level var(), I think init() is consistent in spirit.

A third option is to provide a pair of MustNewXXXX() functions and move these into var(), but then we're back in the business of 2 methods for 3 variables.  I'm not sure I like that any better, though it gives us singleton guarantees as well.

Moving into init() has large overlap with the effort to move into a struct, so unless you specifically want to stay away from init() or there's something I'm missing in that pattern, I can try the init() implementation first.  Let me know thoughts here.

>Also, how would multiple instance effect the recording of these stats?

Just so I'm clear, do you mean calling metricRegistry.AddXXXX concurrently?  

A casual glance looks like they mutex on insertion by name.  So multiple calls with the same name should(?) result in only a single metric with the "last" description and labelkey set.  Type conflicts result in errors (I think, we can test that!).  When the metric is pulled out, it isn't immediately clear that you're getting the same reference.  I could see a race condition where you record to a variable that is later thrown away.

Reply:
> I considered a struct for DRYness, but discounted it. It didn't seem like it would become any less verbose with the current metric count. We have two type signatures and three methods.
> Good point on access safety. These are currently early single threaded calls, but that isn't guaranteed later. This implementation is not safe.
> 
> A lighter pattern might be to do the setup in init(), which I believe is threadsafe, though possibly controversial. Given the rest of the "metric" type variables are initialized in package level var(), I think init() is consistent in spirit.

My thinking on a struct was not so much about DRYness but rather as a means of providing a thread-safe abstraction mutating those package scoped variables (mutex + accessor methods) and pattern for first-time initialization of metric gauges. 

So for example where we have logic like [this](https://github.com/pomerium/pomerium/blob/e0c60d64a7f283524728e3cd287463dbdf175973/internal/metrics/info.go#L53-L62) and [this](https://github.com/pomerium/pomerium/blob/e0c60d64a7f283524728e3cd287463dbdf175973/internal/metrics/info.go#L111-L120) and [this](https://github.com/pomerium/pomerium/blob/e0c60d64a7f283524728e3cd287463dbdf175973/internal/metrics/info.go#L134-L144) we could instead use a singleton and a [DoOnce](https://golang.org/pkg/sync/#Once.Do). What do you think?

Reply:
> A casual glance looks like they mutex on insertion by name. So multiple calls with the same name should(?) result in only a single metric with the "last" description and labelkey set. Type conflicts result in errors (I think, we can test that!). When the metric is pulled out, it isn't immediately clear that you're getting the same reference. I could see a race condition where you record to a variable that is later thrown away.

👍

Reply:
So just a summary - yes I am comfortable that the accessors/recorders (1) are safe but initialization (2) we should guarantee ourselves.  I'll see if the init() pattern looks clean.

Will also see about a test for failing that initialization with a conflicting type.

Reply:
Sounds good to me. 👍

Reply:
I have a [commit](https://github.com/pomerium/pomerium/pull/227/commits/e294188a9f1c7744c9642838748f0294f6539fcb) with the init() style implementation.  It is less wrapping that a struct, but the special handling started to feel a bit brittle and spread out.  The current commit represents a struct based.

I think having the struct container (using the sync.Once pattern) makes it easier to maintain and keep the code behavior in your head.  It also keeps the var() initialization pattern of other data structures.

Additional note - I've intentionally retained the external API, so that callers don't need to be aware of another struct to keep around or reference and we can refactor without much worry.